core/src/components/modal/gestures/sheet.ts

```diff
@@ -294,66 +294,69 @@ export const createSheetGesture = (
-                wrapperAnimation.keyframes([...SheetDefaults.WRAPPER_KEYFRAMES]);
-                backdropAnimation.keyframes([...SheetDefaults.BACKDROP_KEYFRAMES]);
-                animation.progressStart(true, 1 - snapToBreakpoint);
-                currentBreakpoint = snapToBreakpoint;
-                onBreakpointChange(currentBreakpoint);
-
-                /**
-                 * If the sheet is fully expanded, we can safely
-                 * enable scrolling again.
-                 */
-                if (contentEl && currentBreakpoint === breakpoints[breakpoints.length - 1]) {
-                  contentEl.scrollY = true;
-                }
-
-                /**
-                 * Backdrop should become enabled
-                 * after the backdropBreakpoint value
-                 */
-                const shouldEnableBackdrop = currentBreakpoint > backdropBreakpoint;
-                if (shouldEnableBackdrop) {
-                  enableBackdrop();
-                } else {
-                  disableBackdrop();
-                }
-
+    return new Promise((resolve) => {
+      animation
+        .onFinish(
+          () => {
+            if (shouldRemainOpen) {
+              /**
+               * Once the snapping animation completes,
+               * we need to reset the animation to go
+               * from 0 to 1 so users can swipe in any direction.
+               * We then set the animation offset to the current
+               * breakpoint so that it starts at the snapped position.
+               */
+              if (wrapperAnimation && backdropAnimation) {
+                raf(() => {
+                  wrapperAnimation.keyframes([...SheetDefaults.WRAPPER_KEYFRAMES]);
+                  backdropAnimation.keyframes([...SheetDefaults.BACKDROP_KEYFRAMES]);
+                  animation.progressStart(true, 1 - snapToBreakpoint);
+                  currentBreakpoint = snapToBreakpoint;
+                  onBreakpointChange(currentBreakpoint);
+
+                  /**
+                   * If the sheet is fully expanded, we can safely
+                   * enable scrolling again.
+                   */
+                  if (contentEl && currentBreakpoint === breakpoints[breakpoints.length - 1]) {
+                    contentEl.scrollY = true;
+                  }
+
+                  /**
+                   * Backdrop should become enabled
+                   * after the backdropBreakpoint value
+                   */
+                  const shouldEnableBackdrop = currentBreakpoint > backdropBreakpoint;
+                  if (shouldEnableBackdrop) {
+                    enableBackdrop();
+                  } else {
+                    disableBackdrop();
+                  }
+
+                  gesture.enable(true);
+                });
+              } else {
                 gesture.enable(true);
-              });
-            } else {
-              gesture.enable(true);
+              }
+              resolve();
```

Comment:
~ Since this `resolve` is outside the `raf` called on L310, it could be hit before all the logic in there is done. I'm admittedly not sure of the practical implications there, but it may be worth calling it at the end of the `raf`, and after L339 to cover that case as well.

Reply:
+1 we can probably call it in the if/else block.

Reply:
I may need to refactor this approach altogether. Once returning a promise wrapping the `onFinish` callback, it causes the test suite for setting the breakpoint to zero and checking if the event fires, to fail.